### PR TITLE
Enable most IntelliJ 'Probable bugs' inspections

### PR DESCRIPTION
**Pull request by leventov**
Enable "Error" severity for most IntelliJ inspections of "Probable bugs" category which don't catch anything at the moment. Inspections, which have a lot of matches in the codebase, are left on "Warning" level, to keep this PR small. The latter inspections should be moved to "Error" level one-by-one in separate PRs.

A few real (but minor) defects are fixed in the code:
 - Fix `IndexSpec.equals()` and `hashCode()` to include `longEncoding`
 - calling toString() on objects of classes which don't override toString()
 - instance fields which should really be static final fields
 - unnecessary use of `|` and `&` instead of `||` and `&&`.

### CODE REVIEW

File: api/src/main/java/io/druid/data/input/impl/TimestampSpec.java [hunk 1]
```diff
@@ -47,12 +47,12 @@
 
   private final String timestampColumn;
   private final String timestampFormat;
-  private final Function<Object, DateTime> timestampConverter;
   // this value should never be set for production data
   private final DateTime missingValue;
+  private final transient Function<Object, DateTime> timestampConverter;
```
Comment (gianm) [inline api/src/main/java/io/druid/data/input/impl/TimestampSpec.java]: I thought transient just marks fields that shouldn't be serialized. But this class isn't Serializable, so what's the point?

File: common/src/main/java/io/druid/timeline/partition/IntegerPartitionChunk.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
  */
 public class IntegerPartitionChunk<T> implements PartitionChunk<T>
 {
-  Comparator<Integer> comparator = Ordering.<Integer>natural().nullsFirst();
+  private static final Comparator<Integer> comparator = Ordering.<Integer>natural().nullsFirst();
```
Comment (gianm) [inline common/src/main/java/io/druid/timeline/partition/IntegerPartitionChunk.java]: static field should be all-caps.

File: common/src/main/java/io/druid/timeline/partition/LinearPartitionChunk.java [hunk 3]
```diff
@@ -25,7 +25,7 @@
 
 public class LinearPartitionChunk <T> implements PartitionChunk<T>
 {
-  Comparator<Integer> comparator = Ordering.<Integer>natural().nullsFirst();
+  private static final Comparator<Integer> comparator = Ordering.<Integer>natural().nullsFirst();
```
Comment (gianm) [inline common/src/main/java/io/druid/timeline/partition/LinearPartitionChunk.java]: static field should be all-caps.

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 4]
```diff
@@ -1349,10 +1349,12 @@ boolean needConversion(int index)
     {
       IntBuffer readOnly = conversions[index].asReadOnlyBuffer();
       readOnly.rewind();
-      for (int i = 0; readOnly.hasRemaining(); i++) {
+      int i = 0;
+      while (readOnly.hasRemaining()) {
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: What prompted this change?

File: processing/src/main/java/io/druid/segment/data/BitmapCompressedIndexedInts.java [hunk 5]
```diff
@@ -32,7 +32,7 @@
  */
 public class BitmapCompressedIndexedInts implements IndexedInts, Comparable<ImmutableBitmap>
 {
-  private static Ordering<ImmutableBitmap> comparator = new Ordering<ImmutableBitmap>()
+  private static final Ordering<ImmutableBitmap> comparator = new Ordering<ImmutableBitmap>()
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/data/BitmapCompressedIndexedInts.java]: comparator should be all-caps.

File: api/src/main/java/io/druid/data/input/impl/TimestampSpec.java (hunk 1 above)
Comment (leventov) [inline api/src/main/java/io/druid/data/input/impl/TimestampSpec.java]: It was for indication that this field is not checked in equals() and hashCode(). But that was a bad way, removed `transient` and added a simple comment.

File: common/src/main/java/io/druid/timeline/partition/IntegerPartitionChunk.java (hunk 2 above)
Comment (leventov) [inline common/src/main/java/io/druid/timeline/partition/IntegerPartitionChunk.java]: Removed field

File: common/src/main/java/io/druid/timeline/partition/LinearPartitionChunk.java (hunk 3 above)
Comment (leventov) [inline common/src/main/java/io/druid/timeline/partition/LinearPartitionChunk.java]: Removed field

File: processing/src/main/java/io/druid/segment/IndexMerger.java (hunk 4 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: There is an inspection that checks that loop variable is checked in condition part of for loop. This occasion was a false positive, and I think it's better to reformat code than add `@SuppressWarnings`

File: processing/src/main/java/io/druid/segment/data/BitmapCompressedIndexedInts.java (hunk 5 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/data/BitmapCompressedIndexedInts.java]: Renamed

File: processing/src/main/java/io/druid/segment/IndexMerger.java (hunk 4 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Got it.

**Comment by gianm** [comment]
Looks good to me, 👍  after CI passes.

**Comment by leventov** [comment]
Decided to use IntelliJ-style `noinspection` comments to suppress IntelliJ-specific inspections, instead of `@SuppressWarnings` on a method, because the scope of suppression is narrower and it's obvious why suppression is added -- the reason is on the next line. It doesn't matter that those comments are "non-standard", because no other tools or IDEs like Eclipse are able to understand the suppression reasons anyway.

File: processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java [hunk 6]
```diff
@@ -506,16 +506,18 @@ public void read(ByteBuffer buffer, ColumnBuilder builder, ColumnConfig columnCo
       )
       {
         switch (version) {
-          case UNCOMPRESSED_MULTI_VALUE:
+          case UNCOMPRESSED_MULTI_VALUE: {
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java]: I don't think we've ever had this rule before, where is it enforced?

File: processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java [hunk 7]
```diff
@@ -286,11 +286,9 @@ public void testNotEqualCacheables()
 
   private static void assertNotEqualsEachOther(List<byte[]> keys)
   {
-    for (byte[] k1 : keys) {
-      for (byte[] k2 : keys) {
-        if (k1 != k2) {
-          assertFalse(Arrays.equals(k1, k2));
-        }
+    for (int i = 0; i < keys.size(); i++) {
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: why this change?

File: processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java [hunk 8]
```diff
@@ -286,11 +286,9 @@ public void testNotEqualCacheables()
...
+      for (int j = i + 1; j < keys.size(); j++) {
+        assertFalse(Arrays.equals(keys.get(i), keys.get(j)));
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: this logic changes if I'm reading it correctly. Previously if k1 and k2 were the same object, this would not fail. But now it will fail.

File: processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java [hunk 9]
```diff
@@ -286,11 +286,9 @@ public void testNotEqualCacheables()
 
   private static void assertNotEqualsEachOther(List<byte[]> keys)
   {
-    for (byte[] k1 : keys) {
-      for (byte[] k2 : keys) {
-        if (k1 != k2) {
-          assertFalse(Arrays.equals(k1, k2));
-        }
+    for (int i = 0; i < keys.size(); i++) {
+      for (int j = i + 1; j < keys.size(); j++) {
```
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: why start from i+1?

Comment (drcrallen) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: oh, same list, I had in my head different lists

File: processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java (hunk 8 above)
Comment (drcrallen) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: was the prior way just a hack to get over the fact that this *was* doing an n^2 search?

File: server/src/main/java/io/druid/segment/realtime/FireHydrant.java [hunk 10]
```diff
@@ -120,7 +120,7 @@ public String toString()
   {
     return "FireHydrant{" +
            "index=" + index +
-           ", queryable=" + adapter +
+           ", queryable=" + adapter.getIdentifier() +
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/FireHydrant.java]: why this change?

Comment (leventov) [inline server/src/main/java/io/druid/segment/realtime/FireHydrant.java]: Because `ReferenceCountingSegment` doesn't override Object.toString(), so FireHydrant's String form looks like `.., queryable=ReferenceCountingSegment@hash, ..` - not useful.

File: processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java (hunk 8 above)
Comment (leventov) [inline processing/src/test/java/io/druid/query/cache/CacheKeyBuilderTest.java]: As far as I understand, the previous version was looking for duplicates in the keys list. Condition `k1 != k2` was filtering "the diagonal" in n^2 search, but it was unreliable. Anyway, I had to change this piece of code because inspection about array reference comparison.

File: processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java (hunk 6 above)
Comment (leventov) [inline processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java]: This change is a workaround for false positive inspection (due to a defect in the inspection itself). The inspection tracks for "false indentation", e. g.
```java
for (int i : array)
  actionOne();
  actionTwo();
```
Because I may think that both actionOne() and actionTwo() are executed in the loop, and I forgot braces.

In this case, inspection thinks that block
```java
           default:
              throw new IAE("Unsupported multi-value version[%s]", version);		              throw new IAE("Unsupported multi-value version[%s]", version);
```
should have deeper indentation, because the previous case doesn't have `break;` in the end. But if fails to see that all brances in if-else chain break the control flow (two returns and one throw).
